(DOCSP-32576) Document new subscribe API for React Native (#3010)
mongodb/docs-realm: source/sdk/react-native/sync-data/flexible-sync.txt

refer to the :ref:`Flexible Sync RQL Limitations
<react-native-flexible-sync-rql-limitations>` section.

.. _react-native-subscribe-api:

Subscribe to Queries
--------------------

.. versionadded:: ``realm@12.0.0``

``realm@12.0.0`` adds experimental APIs that subscribe to and unsubscribe
from a query's results. These APIs abstract away the details of
manually adding and removing subscriptions.

For all subscriptions, you need :ref:`an authenticated user 
<react-native-authenticate-users>` and a :ref:`Flexible Sync realm 
<react-native-synced-realm>`.

Subscribe to a Query
~~~~~~~~~~~~~~~~~~~~

We recommend that you name your subscriptions. This makes finding and managing
your subscriptions easier. Subscription names must be unique. Trying to add a
subscription with the same name as an existing subscription throws an error.

To subscribe to a query:

> #. Query for the objects that you want to read and write.

Author: The MongoDB style guide suggests not including links in procedure sections, as they distract from what the user is trying to do. https://www.mongodb.com/docs/meta/style-guide/style/links-and-cross-references/#avoid-inline-references-in-tutorials

I'll think about how to include these links somewhere though. I do think they'd be useful.

> #. Query for the objects that you want to read and write.

Author: I've added an "API Reference" sections to the end of procedure topics where needed. The idea is to collect links to the API docs there so that they don't distract readers while they're trying to do something.

> #. Query for the objects that you want to read and write.

Reviewer: Great idea!

   ``subscribe()``.

.. include:: /examples/generated/react-native/v12/BasicSubscription.snippet.subscribe-query.tsx.rst

Most of the time, you should give your subscriptions a name. If you don't, the
name is set to :mdn:`null <Web/JavaScript/Reference/Global_Objects/null>`.

If you use ``filtered()`` on an unnamed query subscription, the subscription
identifier is based on the ``filtered`` query. This means that every time your
query string changes, ``subscribe()`` will create a new subscription.

Wait for a Query Subscription to Sync
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When you subscribe to a query's results, the results do not contain objects
until synced data is downloaded. When you do need to wait for syned objects to
finish downloading, use ``waitForSync``. You can specify different behavior for
your subscriptions and how they handle witing for downloads.

This example uses the ``FirstTime`` option, which is the default behavior.
A subscription with ``FirstTime`` behavior only waits for sync to finish when a
subscription is first created.

.. include:: /examples/generated/react-native/v12/WaitFirstTime.snippet.wait-firsttime.tsx.rst

> The other supported ``WaitForSync`` options are:

Reviewer: Suggestion: Might also be nice to point to the API reference here once for WaitForSync.

When you call ``unsubscribe()``, the associated subscription is removed. 
Subscriptions are removed by name. If they don't have a name, ``unsubscribe()``
removes any queries that exactly match the one you call ``unsubscribe()`` on.

The ``unsubscribe()`` method returns before objects matching the removed
subscription are deleted from the realm. Sync continues in the background based
on the new set of subscriptions.

.. include:: /examples/generated/react-native/v12/Unsubscribe.snippet.unsubscribe-query.tsx.rst

.. _react-native-sync-subscribe-to-queryable-fields:

Manually Manage Subscriptions
-----------------------------

You can use the :js-sdk:`Subscriptions <Realm.App.Sync.Subscription.html>` API
to manually manage a set of subscriptions to specific queries on queryable
fields.

You can use the :realm-react-sdk:`Realm.subscriptions 
<classes/Realm-1.html#subscriptions>` API to manually manage a set of
subscriptions to specific queries on queryable fields.

If you're using ``@realm/react``, you can manage realm subscriptions inside of
a :ref:`properly-configured RealmProvider <react-native-open-a-synced-realm>`. 

> The ``useRealm()`` hook gives you access to the currently-opened realm.

Reviewer: Suggestion: Include useRealm API reference

refer to the :ref:`Flexible Sync RQL Limitations
<react-native-flexible-sync-rql-limitations>` section.

.. _react-native-subscribe-api:

Subscribe to Queries
--------------------

.. versionadded:: ``realm@12.0.0``

``realm@12.0.0`` adds experimental APIs that subscribe to and unsubscribe
from a query's results. These APIs abstract away the details of
manually adding and removing subscriptions.

For all subscriptions, you need :ref:`an authenticated user 
<react-native-authenticate-users>` and a :ref:`Flexible Sync realm 
<react-native-synced-realm>`.

Subscribe to a Query
~~~~~~~~~~~~~~~~~~~~

We recommend that you name your subscriptions. This makes finding and managing
your subscriptions easier. Subscription names must be unique. Trying to add a

> subscription with the same name as an existing subscription throws an error.

Author: Oh, good to know. Not sure why I missed that. Thank you!

To subscribe to a query:

#. Query for the objects that you want to read and write.
#. Call ``subscribe()`` on the query results to create a sync subscription for
   objects matching the query.
#. Pass a ``SubscriptionOptions`` object that contains the ``name`` property to
   ``subscribe()``.

.. include:: /examples/generated/react-native/v12/BasicSubscription.snippet.subscribe-query.tsx.rst

Most of the time, you should give your subscriptions a name. If you don't, the
name is set to ``null``.

If you use ``filtered()`` on an unnamed query subscription, the subscription
identifier is based on the ``filtered`` query. This means that every time your
query string changes, ``subscribe()`` will create a new subscription.

> **API Reference**

Reviewer: Top two links lead to 403 errors.

- :js-sdk:`subscribe() <classes/Results.html#subscribe>`
- :js-sdk:`SubscriptionOptions <types/SubscriptionOptions.html>`
- :mdn:`null <Web/JavaScript/Reference/Global_Objects/null>`

> Wait for a Query Subscription to Sync

Reviewer 2: Been giving this thought, I don't think theres much to change, but I'm trying to think of a situation in React Native where I actually would want to wait for this.  I guess one would want to set some state in this case and block the UI from rendering.  Definitely increases the need of a new hook from `@realm/react`, where the status of the subscription is sent back as a result.

Author: Yeah, I don't think it would be a particularly common case. Ideally, people use initial subs and then pass filtered data to components as needed. But I can imagine some people having questions and thought it might be good to have this. Especially as the Node.js SDK and other SDKs use this syntax (or close to it).

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When you subscribe to a query's results, the results do not contain objects

> until synced data is downloaded. When you do need to wait for syned objects to

Reviewer 2: ```suggestion
until synced data is downloaded. When you do need to wait for synced objects to
```

> finish downloading, use ``waitForSync``. You can specify different behavior for

Reviewer 2: ```suggestion
finish downloading, configure the ``waitForSync`` option. You can specify different behavior for
```

> your subscriptions and how they handle witing for downloads.

Reviewer 2: ```suggestion
your subscriptions and how they handle waiting for downloads.
```

Author: Thanks for catching a couple of little typos!

This example uses the ``FirstTime`` option, which is the default behavior.
A subscription with ``FirstTime`` behavior only waits for sync to finish when a
subscription is first created.

.. include:: /examples/generated/react-native/v12/WaitFirstTime.snippet.wait-firsttime.tsx.rst

The other supported ``WaitForSync`` options are:

- ``Always``: Wait to download matching objects every time your app launches.
  The app must have an internet connection at every launch.
- ``Never``: Never wait to download matching objects. The app needs an internet
  connection for the user to authenticate the first time the app launches, but
  can :ref:`open offline <react-native-open-synced-realm-offline>` on
  subsequent launches using cached credentials.

You can optionally specify a ``timeout`` value to limit how long the sync
download runs:

.. include:: /examples/generated/react-native/v12/AlwaysWait.snippet.always-sync.tsx.rst

**API Reference**

> - :js-sdk:`waitForSync </enums/WaitForSync.html>`

Reviewer: 403 error

> Unsubscribe from a Query

Reviewer 2: Could be worth noting, subscriptions that have been added will persist upon the app opening again, unless the subscription has been removed.

~~~~~~~~~~~~~~~~~~~~~~~~

You can unsubscribe from a query's results using ``unsubscribe()``:

This removes the subscription from the list of active subscriptions, similar to
:ref:`manually removing a subscription <react-native-remove-subscription>`.

A results list may still contain objects after calling ``unsubscribe()`` 
if another subscription exists that contains overlapping objects.

When you call ``unsubscribe()``, the associated subscription is removed. 
Subscriptions are removed by name. If they don't have a name, ``unsubscribe()``
removes any queries that exactly match the one you call ``unsubscribe()`` on.

The ``unsubscribe()`` method returns before objects matching the removed
subscription are deleted from the realm. Sync continues in the background based
on the new set of subscriptions.

.. include:: /examples/generated/react-native/v12/Unsubscribe.snippet.unsubscribe-query.tsx.rst

**API Reference**

> - :js-sdk:`unsubscribe() </classes/Results.html#unsubscribe>`

Reviewer: 403 error

When the data matches the subscription, and the authenticated user has the
appropriate permissions, Device Sync syncs the backend data with the client app.

When you create a subscription, Realm looks for data matching a query on a
specific object type. You can have subscriptions on several different object
types. You can also have multiple queries on the same object type.

.. important:: Object Links

   You must add both an object and its linked object to the subscription 
   set to see a linked object.
   
   If your subscription results contain an object with a property that links 
   to an object not contained in the results, the link appears to be null.
   There is no way to distinguish whether that property's value is 
   legitimately null, or whether the object it links to exists but is out of
   view of the query subscription.

.. _react-native-get-subscriptions:

Access All Subscriptions
~~~~~~~~~~~~~~~~~~~~~~~~

Within a ``RealmProvider`` that's configured for Flexible Sync, you can access
a collection of all subscriptions for your app. This is called a

> ``SubscriptionSet``.

Reviewer: Suggestion: Could use API reference for SubscriptionSet here.

When you create a subscription, Realm looks for data matching a query on a
specific object type. You can have subscriptions on several different object
types. You can also have multiple queries on the same object type.

.. important:: Object Links

   You must add both an object and its linked object to the subscription 
   set to see a linked object.
   
   If your subscription results contain an object with a property that links 
   to an object not contained in the results, the link appears to be null.
   There is no way to distinguish whether that property's value is 
   legitimately null, or whether the object it links to exists but is out of
   view of the query subscription.

.. _react-native-get-subscriptions:

Access All Subscriptions
~~~~~~~~~~~~~~~~~~~~~~~~

Within a ``RealmProvider`` configured for Flexible Sync, you can access
a ``SubscriptionSet``. A SubscriptionSet is a collection of all subscriptions
for your app.

.. literalinclude:: /examples/generated/react-native/ts/get-subscriptions.test.snippet.get-subscriptions.tsx
   :language: javascript
   :emphasize-lines: 12

**API Reference**

> - :js-sdk:`SubscriptionSet </classes/SubscriptionSet.html>`

Reviewer: 403 Error

The Realm SDK does not provide a way to check whether all documents that match 
a subscription have synced to the device.

Update Subscriptions with a New Query
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can update a named subscription with a new query. To update a subscription's
query, pass the new query and a subscription option with the name of the

> subscription that you want to update to the ``MutableSubscriptionSet.add()``

Reviewer: Suggestion: API reference for .add()

method. Like adding a new subscription, you must update a subscription within a

> transaction by calling ``subscriptions.update()``.

Reviewer: Suggestion: API reference for .update()

In the following example, long-running tasks are re-defined to be any tasks that
have taken more than 180 minutes. 

.. literalinclude:: /examples/generated/node/flexible-sync.snippet.update-subscriptions.js

.. note::

   Attempting to update a subscription that has the
   ``SubscriptionOptions.throwOnUpdate`` field set to true throws an exception.

.. _react-native-remove-subscription:

Remove Subscriptions
~~~~~~~~~~~~~~~~~~~~

Subscription sets persist across sessions, even if you no longer include
the subscription in your code. Subscription information is stored in the synced
realm's database file. You must explicitly remove a subscription for it to stop
attempting to sync matching data.

> You can remove subscriptions in several ways:

Reviewer: ```suggestion
You can remove subscriptions in the following ways:
```
Word change: Are these the only ways to remove the subscription? 'Several' to me makes it seem like there could be others.

The Realm SDK does not provide a way to check whether all documents that match 
a subscription have synced to the device.

Update Subscriptions with a New Query
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can update a named subscription with a new query. To update a subscription's
query, pass the new query and a subscription option with the name of the
subscription that you want to update to the ``MutableSubscriptionSet.add()``
method. Like adding a new subscription, you must update a subscription within a
transaction by calling ``subscriptions.update()``.

The following example, redefines long-running tasks as any tasks that take
more than 180 minutes.

.. literalinclude:: /examples/generated/node/flexible-sync.snippet.update-subscriptions.js

.. note::

   Attempting to update a subscription that has the
   ``SubscriptionOptions.throwOnUpdate`` field set to true throws an exception.

> **API Reference**

Reviewer: Links need to be fixed, these API reference links are leading to pages with 403 errors.

Author: Thanks for catching this too! Did platform change how the role links behave. I'm pretty sure I needed a preceding `/` before, but now those break the URLs...

Reviewer: I'm honestly not sure! I just looked at the [:js-sdk: defintion](https://github.com/mongodb/snooty-parser/blob/095c4d2ad8fba25ba79406f643de0304e82abc42/snooty/rstspec.toml#L1458), and there is no ending ``/``, so perhaps :(

<Realm.App.Sync.MutableSubscriptionSet.html#removeByObjectType>`.

.. literalinclude:: /examples/generated/node/flexible-sync.snippet.remove-all-subscriptions-of-object-type.js
   :language: javascript

Remove All Unnamed Subscriptions
`````````````````````````````````

.. versionadded:: ``realm@v12.0.0``

You may want to remove unnamed subscriptions that are transient or dynamically 
generated, but leave named subscriptions in place.

You can remove all unnamed subscriptions from the subscription set by 

> calling ``.removeUnnamed()`` on ``mutableSubs``. ``.removeUnnamed()`` returns

Reviewer: API ref for .removeUnnamed()?

<Realm.App.Sync.MutableSubscriptionSet.html#removeByObjectType>`.

.. literalinclude:: /examples/generated/node/flexible-sync.snippet.remove-all-subscriptions-of-object-type.js
   :language: javascript

Remove All Unnamed Subscriptions
`````````````````````````````````

.. versionadded:: ``realm@v12.0.0``

You may want to remove unnamed subscriptions that are transient or dynamically 
generated, but leave named subscriptions in place.

You can remove all unnamed subscriptions from the subscription set by 
calling ``.removeUnnamed()`` on ``mutableSubs``. ``.removeUnnamed()`` returns
the number of unnamed subscriptions removed.

.. literalinclude:: /examples/generated/node/v12/manage-subscriptions.test.snippet.sub-remove-unnamed.ts

**API Reference**

> - :js-sdk:`MutableSubscriptionSet.add() </classes/MutableSubscriptionSet.html#removeUnnamed>`

Reviewer: Is this .add() reference necessary here in the remove section?

Author: Oops. Thank you for catching that! The URL is correct, but I forgot to change the display text. Definitely running low on energy this afternoon!
